kubric/datasets/klevr.py

```diff
@@ -0,0 +1,176 @@
+  "source": tf.string,
+  "RGBA": tfds.features.Tensor(shape=(H, W, 4), dtype=tf.uint8),
+  "segmentation": tfds.features.Tensor(shape=(H, W, 1), dtype=tf.uint32),
+  "flow": tfds.features.Tensor(shape=(H, W, 3), dtype=tf.float32),
+  "depth": tfds.features.Tensor(shape=(H, W, 1), dtype=tf.float32),
+  "UV": tfds.features.Tensor(shape=(H, W, 3), dtype=tf.float32)
+})
+
+# ------------------------------------------------------------------------------
+# ------------------------------------------------------------------------------
+# ------------------------------------------------------------------------------
+
+class Klevr(tfds.core.GeneratorBasedBuilder):
+  VERSION = VERSION
+
+  def __init__(self, path, train_to_test_ratio=.7, *args, **kwargs):
+    super().__init__(*args, **kwargs)
+    logging.info(f"Klevr(path={path})")
+    self.path = path
+
+    # --- open metadata
+    with tf.io.gfile.GFile(str(self.path + "/metadata.pkl"), "rb") as fp:
+      self.metadata = pickle.load(fp)
+
+    # --- split information
+    self.n_frames = len(self.metadata)
+    self.n_train = math.floor(self.n_frames * train_to_test_ratio)
+    self.n_test = self.n_frames - self.n_train
+
+  def _info(self):
+    return tfds.core.DatasetInfo(
+      builder=self,
+      description=DESCRIPTION,
+      features=FEATURES,
+      citation=CITATION)
+
+  def _split_generators(self, _):
+    # --- split information
+    shuffled_ids = np.random.permutation(np.arange(self.n_frames+1)[1:])
+    ids_train = shuffled_ids[0:self.n_train]
+    ids_test = shuffled_ids[self.n_train:]
+
+    # --- instantiate splits (→_generate_examples)
+    return {
+      "train": self._generate_examples(ids_train),
+      "test": self._generate_examples(ids_test) 
+    }
+
+  def _generate_examples(self, ids):
+    for frame_id in ids:
+      frame_path = self.path + f"/frame_{frame_id:04d}.pkl"
+
+      with tf.io.gfile.GFile(frame_path, "rb") as fp:
+        data = pickle.load(fp)
+
+        yield frame_path, {
+          "source": frame_path,
+          "RGBA": data["RGBA"],
+          "segmentation": data["segmentation"],
+          "flow": data["flow"],
+          "depth": data["depth"],
+          "UV": data["UV"]
+        }
+    
+# ------------------------------------------------------------------------------
+# ------------------------------------------------------------------------------
+# ------------------------------------------------------------------------------
+
+class MasterKlevr(tfds.core.GeneratorBasedBuilder):
+  VERSION = VERSION
+
+  def __init__(self, path, *args, **kwargs):
+    super().__init__(*args, **kwargs)
+    
+    # --- create a builder for each subfolder
+    self._builders = list()
+    subdirs = tf.io.gfile.listdir(path)
+    # removes trailing "/"" caused by listdir, and recovers full path
+    subdirs = [path+"/"+subdir[:-1] for subdir in subdirs]
+
+    # --- generate the dataset builders 
+    for isubdir, subdir in enumerate(subdirs):
```

Comment:
You likely want to move this code inside _split_generators, as this is only required during generation. Otherwise this will add overhead for all reader